enhancements/compound-bundle-constraints.md

```diff
@@ -0,0 +1,378 @@
+
+As a cluster administrator, I should be able to opt-in to resolving compound
+constraints in OLM while the feature is alpha. When opted-out, the resolver
+should emit errors so bundles are not installed incorrectly, ex. missing an API.
+
+### Implementation Details/Notes/Constraints
+
+**Note**: see ["Implementation History"](#implementation-history) for proof-of-concept info.
+
+I propose the new properties `olm.all.required`, `olm.any.required`, and `olm.none.required`
+to specify conjunction, disjunction, and negation compound constraints, respectively.
+Their value type is a struct containing a `description`, and list of `olm.package.required`,
+`olm.gvk.required`, or any compound constraint listed above under a `constraints` key.
+
+The implementation of these constraints will primarily live in the OLM codebase,
+since they are properties that the resolver understands. Some parsing details
+will be implemented in operator-registry for registry components that need to know
+about dependencies.
+
+#### Conjunction and disjunction
+
+These compound constraint types are evaluated following their logical definitions.
+
+This is an example of a conjunctive constraint of two packages and one GVK,
+i.e. they must all be satisfied by installed bundles:
+
+```yaml
+schema: olm.bundle
+name: baz.v1.0.0
+properties:
+- type: olm.all.required
+  value:
+    description: All are required for Baz because...
+    constraints:
+    - type: olm.package.required
+      value:
+        packageName: baz
+        versionRange: '>=1.0.0'
+    - type: olm.package.required
+      value:
+        packageName: buf
+        versionRange: '>=1.0.0'
+    - type: olm.gvk.required
+      value:
+        group: bazs.example.com
+        version: v1
+        kind: Baz
+```
+
+This is an example of a disjunctive constraint of three versions of the same GVK,
+i.e. at least one must be satisfied by installed bundles:
+
+```yaml
+schema: olm.bundle
+name: baz.v1.0.0
+properties:
+- type: olm.any.required
+  value:
+    description: Any are required for Baz because...
+    constraints:
+    - type: olm.gvk.required
+      value:
+        group: foos.example.com
+        version: v1beta1
+        kind: Foo
+    - type: olm.gvk.required
+      value:
+        group: foos.example.com
+        version: v1beta2
+        kind: Foo
+    - type: olm.gvk.required
+      value:
+        group: foos.example.com
+        version: v1
+        kind: Foo
+```
+
+#### Negation
+
+Negation is worth further explanation, since at first glance its semantics
+are unclear in this context. The negation is really instructing the resolver
+to remove any possible solution that includes a particular GVK, package
+at a version, or satisfies some child compound constraint from the result set.
+
+This is an example of a negation constraint of one version of a GVK,
+i.e. this GVK cannot satisfy the constraint:
+
+```yaml
+schema: olm.bundle
+name: baz.v1.0.0
+properties:
+- type: olm.none.required
```

Comment:
this is interesting, but I think it more matches "conflicts" constraints in the resolver than it does negation.

negation would let you say things like `A && B || A && !C` etc

Reply:
I am unclear on what you mean. Negation here intentionally means "use any GVK/package that satisfies other constraints except these GVKs/packages". Are you saying this is not useful, or just that the implementation of negation in the resolver is "conflicts"?